### ISSUE TITLE
"given value instantiates a different structure type with the same name" when printing

### ISSUE THREAD

**Issue by jsmaniac** [racket/racket#1579]
When running this file:

```racket
#lang racket
(require (for-syntax racket
                     racket/struct
                     racket/pretty))

(begin-for-syntax
  (define ((struct-printer ctor) st port mode)
    (fprintf port "(~a " ctor)
    (for-each (λ (field)
                (display (with-output-to-string
                          (λ ()
                            (print field (current-output-port) 0)))
                         port))
              (struct->list st))
    (fprintf port ")"))

  (struct the-struct (predicate)
    #:transparent
    #:methods gen:custom-write
    [(define write-proc (struct-printer 'invariant-info))]
    #:property prop:custom-print-quotable 'never)
  (pretty-print (the-struct (list (make-string 100 #\x)
                                  (set (the-struct 'c))))))
```

I get the following error when running with <kbd>F5</kbd> in DrRacket:
```
../../../../../../nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/pretty.rkt:747:2: printing-port-info: contract violation;
 given value instantiates a different structure type with the same name
  expected: printing-port?
  given: #<printing-port>
```

and a slightly different error when running on the command-line:
```
display/recursive: contract violation
  expected: output-port?
  given: #<printing-port>
  argument position: 2nd
  other arguments...:
   "set"
  context...:
   /nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/private/custom-write.rkt:60:12
   /nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/private/custom-write.rkt:9:2
   /nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/port.rkt:125:0: with-output-to-string
   /nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/private/map.rkt:57:19: loop
   /tmp/bb.rkt:7:2
   /nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/pretty.rkt:1013:4: pr
   /nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/pretty.rkt:414:0: generic-write17
   /nix/store/6gd54zj6yp7q4b3mp6dzmlb0c1yg6frm-racket-6.8.0.1/share/racket/collects/racket/pretty.rkt:212:14: unpack37
   standard-module-name-resolver
```

Here are some relevant points:
* My custom printer tries to circumvent `print`'s undocumented auto-detection which switches to quoted form (e.g. `'#hash((k . v) …)` instead of `(hash k v …)`) in some cases.
* To this end, I'm printing the fields of the struct to intermediate strings, using `with-output-to-string`, and then print those strings to the port:
* If the code is moved out of the `(begin-for-syntax …)`, it works
* The above works fine, except when one of the fields of the struct is a `set`
* The error seems to be sensitive to the line width: if I reduce the length of the `(make-string 100 #\x)` string, it works fine
* The error seems to be triggered only when wrapping `(the-struct 'c)` on the last line with a `set`, otherwise it prints fine.

**Comment by mflatt** [comment]
I think the problem in DrRacket is that the output port spans phases, and it uses a `racket/pretty` at phase 0 relative to the program running in DrRacket, so it doesn't cooperate nicely with `racket/pretty` at phase 1. Probably something in `racket/pretty` is incorrectly assuming that it gets a compatible port via callbacks, and I'll investigate more.

The problem in Racket is a bug in `gen:custom-write` handling, and I'll push a repair for that soon.

Meanwhile, I think the way that `print` switches to quoted form is defined in "1.4 The Printer" in the Reference (where it talks about quoting depth). Does that not cover what you had in mind?

**Comment by jsmaniac** [comment]
@mflatt Thanks for looking into this!

Concerning the docs, they explain what causes print to switch to the quoted form, but there seems to be no `parameter` to control that (i.e. prevent switching to quoted form when the contents are quotable). I had found out that in my particular case, printing to a temporary string and then printing that string prevented that behaviour, for some reason. In the end, though, I switched to using `mzlib/pconvert`.
